src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,236 @@
+import "@polymer/paper-fab/paper-fab.js";
+
+import "../../../components/ha-card.js";
+
+import { hassLocalizeLitMixin } from "../../../mixins/lit-localize-mixin";
+import { HomeAssistant } from "../../../types.js";
+import { LovelaceCard, LovelaceConfig } from "../types.js";
+import {
+  clearItems,
+  fetchItems,
+  addItem,
+  completeItem,
+  saveEdit,
+  deleteItem,
+} from "../../../data/shopping_list";
+import { TemplateResult } from "lit-html";
+
+interface Config extends LovelaceConfig {
+  title?: string;
+}
+
+export interface ShoppingListItem {
+  id: number;
+  name: string;
+  complete: boolean;
+}
+
+class HuiShoppingListCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  protected _config?: Config;
+  private _items?: ShoppingListItem[];
+  private _unsubEvents?: () => Promise<void>;
+
+  static get properties() {
+    return {
+      _config: {},
+      _items: {},
+    };
+  }
+
+  public getCardSize(): number {
+    return (
+      (this._config ? (this._config.title ? 1 : 0) : 0) +
+      (this._items ? this._items.length : 3)
+    );
+  }
+
+  public setConfig(config: Config): void {
+    this._config = config;
+    this._items = [];
+    this._fetchData();
+  }
+
+  public connectedCallback() {
+    super.connectedCallback();
+
+    if (this.hass) {
+      this.hass.connection
+        .subscribeEvents(() => this._fetchData(), "shopping_list_updated")
+        .then(
+          (unsub: () => Promise<void>): void => {
+            this._unsubEvents = unsub;
+          }
+        );
+      this._fetchData();
+    }
+  }
+
+  public disconnectedCallback() {
+    super.disconnectedCallback();
+
+    if (this._unsubEvents) {
+      this._unsubEvents();
+    }
+  }
+
+  protected render(): TemplateResult {
+    if (!this._config || !this.hass) {
+      return html``;
+    }
+
+    return html`
+      ${this.renderStyle()}
+      <ha-card .header="${this._config.title}">
+        <div class="addRow">
+          <paper-input
+            id='addBox'
+            placeholder='${this.localize("ui.card.shopping-list.add_item")}'
+            @keydown=${this._addKeyPress}
+            no-label-float
+          ></paper-input>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
+            @click=${this._addItem}
+          ></paper-icon-button>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:notification-clear-all"
+            @click=${this._clearItems}
+          ></paper-icon-button>
+        </div>
+
+      ${repeat(
+        this._items!,
+        (item) => html`
+        <div class="editRow">
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
+            .itemId=${item.id}
+            @click=${this._completeItem}
+            tabindex='0'
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input

```

Comment:
Look, adding this is not MVP.

MVP stands for minimum viable product. The least amount of logic to get this card merged. So that would just be a card that renders items and allows checkoxes.

Instead this PR tries to do EVERYTHING . Adding items, clearing items, checking items, editing individual items.

Reply:
I get MVP, but really don't like that idea. I know this has been quite rough going through this; so maybe I'll take a step back and make sure this is truly good to go before making any more commits or perhaps even removing this PR until I have polish.

Reply:
You might not like the idea of MVP, but then there are multiple people involved in a PR. 

MVP makes everyones life's easier.

- Smallish initial PR to get basic functionality in place. Easy to see why code is there.
- New PR to add feature X, we see exactly each line that is related to feature X.

Reply:
Just to be clear: I am not reviewing this PR any further for now. There are over 250 open PRs in our org right now, many of which are older than this one. Many which are focusing on a single feature or MVP.